src/main/java/com/google/sps/data/Cache.java

```diff
@@ -1,90 +1,77 @@
     }
-
-    Coordinates nwCorner = new Coordinates(swCorner.lng, neCorner.lat);
-    Coordinates seCorner = new Coordinates(neCorner.lng, swCorner.lat);
-    int resolution = 100000;
-    GridIndex lowerRightIndex = getGridIndex(nwCorner, seCorner, resolution);
-    int gridWidth = lowerRightIndex.col + 1;
-    int gridHeight = lowerRightIndex.row + 1;
-    GridCell[][] processingGrid = new GridCell[gridHeight][gridWidth];
-
+    
     for (AQDataPoint dataPoint : data) {
-        GridIndex index = getGridIndex(nwCorner, new Coordinates(dataPoint.lng, dataPoint.lat),resolution);
-        if (index.row >= gridHeight || index.col >= gridWidth) {
-            // The data point is outside the area covered by the grid
-            continue;
-        }
-
-        if (processingGrid[index.row][index.col] == null) {
-            processingGrid[index.row][index.col] = new GridCell();
-        }
-        processingGrid[index.row][index.col].addPoint(dataPoint);
+      GridIndex index = getGridIndex(new Coordinates(dataPoint.lng, dataPoint.lat), aqDataPointsPerDegree);
+      HashMap<Integer,GridCell> row = dataGrid.getOrDefault(index.row, new HashMap<>());
+      GridCell cell = row.getOrDefault(index.col, new GridCell());
+      cell.addPoint(dataPoint);
+      row.put(index.col, cell);
+      dataGrid.put(index.row, row);
     }
 
-    dataGrid = new GriddedData(processingGrid, resolution, nwCorner);
-    return dataGrid;
+    return convertToGriddedData(dataGrid, aqDataPointsPerDegree, swCorner, neCorner);
   }
 
   private ArrayList<AQDataPoint> getNSWGovData() throws Exception {
     NSWGovAdaptor adaptor = new NSWGovAdaptor();
     return adaptor.getAQIData();
   }
 
-  private GridIndex getGridIndex(Coordinates originCoords, Coordinates targetCoords, int resolution) {
-    /*
-    * Layout of coordinates:
-    * originCoords -------------------- sameLngCoords
-    *       |                                  |
-    *       |                                  |
-    *       |                                  |
-    * sameLatCoords -------------------- targetCoords
-    * 
-    * Note: Due to the curvature of the earth, the actual shape bounded by the four
-    * points may not be exactly rectangular, however it should be close enough
-    * for small distances.
-    */
-    Coordinates sameLatCoords = new Coordinates(originCoords.lng,targetCoords.lat);
-    Coordinates sameLngCoords = new Coordinates(targetCoords.lng,originCoords.lat);
-
-    double latDistance = haversineDistance(originCoords, sameLngCoords);
-    double lngDistance = haversineDistance(originCoords, sameLatCoords);
-
-    int col = (int) Math.floor(latDistance / resolution);
-    int row = (int) Math.floor(lngDistance / resolution);
-
+  private GridIndex getGridIndex(Coordinates targetCoords, int aqDataPointsPerDegree) {
+    int row = (int) Math.floor(targetCoords.lat * aqDataPointsPerDegree);
```

Comment:
```suggestion
    int row = (int) (targetCoords.lat * aqDataPointsPerDegree);
```

The `(int)` cast already discards the floating component.